### ISSUE TITLE
VersionsAndSeqNoResolver may experience delayed ThreadLocal cleanup with high pressure writing to ES

### ISSUE THREAD

**Issue by hydrogen666** [elastic/elasticsearch#56766]
We have several Elasticsearch cluster running Elasticsearch v6.2.4, when with high pressure writing to Elasticsearch, some data node may got bulk thread pool full issue, hot thread:
```text
::: {xxxxxxxxx}{lCj7LcVnT328KHcJRd57yg}{WPiNCbk0R0SIKxg4-w3wew}{xxxxxxxx}{xxxxxxxx}
   Hot threads at 2020-04-12T05:25:10.224Z, interval=500ms, busiestThreads=3, ignoreIdleThreads=true:
   
   105.3% (526.5ms out of 500ms) cpu usage by thread 'elasticsearch[xxxxxxxx][bulk][T#31]'
     10/10 snapshots sharing following 34 elements
       java.lang.ThreadLocal$ThreadLocalMap.expungeStaleEntry(ThreadLocal.java:627)
       java.lang.ThreadLocal$ThreadLocalMap.remove(ThreadLocal.java:509)
       java.lang.ThreadLocal$ThreadLocalMap.access$200(ThreadLocal.java:308)
       java.lang.ThreadLocal.remove(ThreadLocal.java:224)
       java.util.concurrent.locks.ReentrantReadWriteLock$Sync.tryReleaseShared(ReentrantReadWriteLock.java:426)
       java.util.concurrent.locks.AbstractQueuedSynchronizer.releaseShared(AbstractQueuedSynchronizer.java:1349)
       java.util.concurrent.locks.ReentrantReadWriteLock$ReadLock.unlock(ReentrantReadWriteLock.java:881)
       org.elasticsearch.common.util.concurrent.ReleasableLock.close(ReleasableLock.java:49)
       org.elasticsearch.index.engine.InternalEngine.$closeResource(InternalEngine.java:356)
       org.elasticsearch.index.engine.InternalEngine.delete(InternalEngine.java:1272)
       org.elasticsearch.index.shard.IndexShard.delete(IndexShard.java:812)
       org.elasticsearch.index.shard.IndexShard.applyDeleteOperation(IndexShard.java:779)
       org.elasticsearch.index.shard.IndexShard.applyDeleteOperationOnReplica(IndexShard.java:750)
       org.elasticsearch.action.bulk.TransportShardBulkAction.performOpOnReplica(TransportShardBulkAction.java:623)
       org.elasticsearch.action.bulk.TransportShardBulkAction.performOnReplica(TransportShardBulkAction.java:577)
```
The thread is doing clean work on useless ThreadLocal$Entry, after analyzing the heap of the process, we got about 10K+ ThreadLocal$Entry for bulk threads with null `referent` field and `WeakReference` value. It indicates `org.apache.lucene.util.CloseableThreadLocal`(which is using `ThreadLocal<WeakReference<T>>`) may be the cause of the leak.

After reading [VersionsAndSeqNoResolver.java](https://github.com/elastic/elasticsearch/blob/master/server/src/main/java/org/elasticsearch/common/lucene/uid/VersionsAndSeqNoResolver.java) of es version 6.2.4 and master, the ThreadLocal is set in [line 75](https://github.com/elastic/elasticsearch/blob/master/server/src/main/java/org/elasticsearch/common/lucene/uid/VersionsAndSeqNoResolver.java#L75) and closed in [line 44](https://github.com/elastic/elasticsearch/blob/master/server/src/main/java/org/elasticsearch/common/lucene/uid/VersionsAndSeqNoResolver.java#L44). The `set` method is called on write thread_pool (bulk thread pool for 6.2.4), and `close` method is called on refresh thread_pool, it means the close method wont clean `ThreadLocal`s on write thread_pool, it may cause ThreadLocal leak. 

see also 
[HBASE-16616](https://issues.apache.org/jira/browse/HBASE-16616)
[JENKINS-58899](https://issues.jenkins-ci.org/browse/JENKINS-58899?page=com.atlassian.streams.streams-jira-plugin%3Aactivity-stream-issue-tab)

**Comment by DaveCTurner** [comment]
I don't see the relationship to `org.apache.lucene.util.CloseableThreadLocal`, the hot threads you shared is busy with the `ThreadLocalMap` that underpins a `java.util.concurrent.locks.ReentrantReadWriteLock`. I also don't think there's a problem with calling `CloseableThreadLocal#close()` on a different thread from the one that created it. That looks like the way it's intended to be used, unless I'm missing something.

Version 6.2.4 is well past the [end of its supported life](https://www.elastic.co/support/eol). Can you reproduce this on a supported version? Can you share your heap dump?

**Comment by hydrogen666** [comment]
> I don't see the relationship to `org.apache.lucene.util.CloseableThreadLocal`, the hot threads you shared is busy with the `ThreadLocalMap` that underpins a `java.util.concurrent.locks.ReentrantReadWriteLock`. I also don't think there's a problem with calling `CloseableThreadLocal#close()` on a different thread from the one that created it. That looks like the way it's intended to be used, unless I'm missing something.
> 
> Version 6.2.4 is well past the [end of its supported life](https://www.elastic.co/support/eol). Can you reproduce this on a supported version? Can you share your heap dump?

We may not be able to share the heapdump because we have sensitive data in the heap dump, but I can do some screen shot on this.
32 ` java.lang.ThreadLocal$ThreadLocalMap`s with entry count > 10k, and we have a 32 thread bulk_thread_pool and these ` java.lang.ThreadLocal$ThreadLocalMap` belongs to the bulk threads
![image](https://user-images.githubusercontent.com/17194883/81942332-958d1500-962c-11ea-9164-78afe4d99fe4.png)

 398k+ `java.lang.ThreadLocal$ThreadLocalMap$Entry`(about 10k+ per bulk thread) with value of type `java.lang.ref.WeakReference`(only `CloseableThreadLocal` do have `ThreadLocal<WeakReference<T>>`) and with null referent(`java.lang.ThreadLocal$ThreadLocalMap$Entry` is a weak reference, its referent will be GCed after GC)
![image](https://user-images.githubusercontent.com/17194883/81942797-29f77780-962d-11ea-9a8d-d19613c99dd4.png)



It does have relationship with `CloseableThreadLocal `, if `set` an object to `CloseableThreadLocal` in thread A(in our example the bulk thread), and `close` it in thread B(in our example the refresh thread), the object `set` in the thread A will not be cleared. Every thread have its own `ThreadLocal$ThreadLocalMap`, if we dont close the `CloseableThreadLocal` explicitly in thread A, it will cause thread local leak, see [HBASE-16616](https://issues.apache.org/jira/browse/HBASE-16616). 

Why we say it does have relationship with `CloseableThreadLocal`, because the `ReentrantReadWriteLock ` is also using the `ThreadLocal` in the bulk thread, which share the same `ThreadLocal$ThreadLocalMap` with `CloseableThreadLocal` in  ` VersionsAndSeqNoResolver`. 

When calling `ReentrantReadWriteLock$ReadLock#unlock` method, it will call `ThreadLocal$ThreadLocalMap#expungeStaleEntry` method, it will try to clean any stale entry(with `ThreadLocal$ThreadLocalMap$Entry#referent` is null).  Because we did not close the `CloseableThreadLocal` in the same thread of calling the `set` method, it will leave a lot of stale `ThreadLocal$ThreadLocalMap$Entry`s in `ThreadLocal$ThreadLocalMap$Entry`. It will cost much cpu to 'expunge' the stale entries.

**Comment by dnhatn** [comment]
@hydrogen666 I looked at the issue closer and I think I understand the problem. I will try to fix it (a bit tricky as we must ensure that we do not degrade the indexing performance). Thank you for reporting it.

**Comment by dnhatn** [comment]
@hydrogen666 FYI, this issue on 7.x is much less severe as we [skip refreshes ](https://www.elastic.co/guide/en/elasticsearch/reference/current/release-highlights-7.0.0.html#_skip_shard_refreshes_if_a_shard_is_search_idle)if a shard is search idle.

**Comment by hydrogen666** [comment]
@dnhatn I think we can use a `ConcurrentHashMap` with `Thread.currentThread()` as key instead of `ThreadLocal`

**Comment by dnhatn** [comment]
@hydrogen666 Yes, I am working on the fix.

**Comment by dnhatn** [comment]
@hydrogen666 Several places in Lucene also use CloseableThreadLocal. I do not want to miss these places. Can you share the heap dump privately so we can confirm these ThreadLocal are really from VersionsAndSeqNoResolver ? Thank you!

**Comment by hydrogen666** [comment]
@dnhatn I need to ask my boss whether I can share this dump with you. How can I get contact with you?

**Comment by hydrogen666** [comment]
@dnhatn Too much sensitive data in this heap dump, we can communicate via IM like DingTalk or Skype if you want any informations.

**Comment by dnhatn** [comment]
@hydrogen666 You can DM me on https://discuss.elastic.co

**Comment by hydrogen666** [comment]
@dnhatn sent u a message on https://discuss.elastic.co

**Comment by hydrogen666** [comment]
Hi @dnhatn, any further information about this issue?

**Comment by cztchoice** [comment]
Hi, do we have any update on this? We face this issues too.

**Comment by dnhatn** [comment]
@cztchoice Can you share the ES and node settings (i.e., GC settings, number of CPUs), and [hot_threads](https://www.elastic.co/guide/en/elasticsearch/reference/current/cluster-nodes-hot-threads.html)? It would be great if you can share a heap dump (privately).

**Comment by cztchoice** [comment]
we use es version 7.4.2 on a 16C64G machine

CPU Architecture
```
Architecture:          x86_64
CPU op-mode(s):        32-bit, 64-bit
Byte Order:            Little Endian
CPU(s):                16
On-line CPU(s) list:   0-15
Thread(s) per core:    1
Core(s) per socket:    16
Socket(s):             1
NUMA node(s):          1
Vendor ID:             GenuineIntel
CPU family:            6
Model:                 94
Model name:            Intel(R) Xeon(R) Gold 6133 CPU @ 2.50GHz
Stepping:              3
CPU MHz:               2494.142
BogoMIPS:              4988.28
Hypervisor vendor:     KVM
Virtualization type:   full
L1d cache:             32K
L1i cache:             32K
L2 cache:              4096K
L3 cache:              28160K
NUMA node0 CPU(s):     0-15
```

java command(gc settings)

```
/usr/java/jdk-11.0.3/bin/java -Xms31g -Xmx31g -XX:+UseG1GC -XX:MaxGCPauseMillis=200 -XX:InitiatingHeapOccupancyPercent=45 -XX:+ParallelRefProcEnabled -XX:+UseBiasedLocking -XX:+UnlockExperimentalVMOptions -XX:G1NewSizePercent=30 -XX:G1HeapRegionSize=16m -XX:MaxMetaspaceSize=512m -Des.networkaddress.cache.ttl=60 -Des.networkaddress.cache.negative.ttl=10 -XX:+AlwaysPreTouch -Xss1m -Djava.awt.headless=true -Dfile.encoding=UTF-8 -Djna.nosys=true -XX:-OmitStackTraceInFastThrow -Dio.netty.noUnsafe=true -Dio.netty.noKeySetOptimization=true -Dio.netty.recycler.maxCapacityPerThread=0 -Dlog4j.shutdownHookEnabled=false -Dlog4j2.disable.jmx=true -Djava.io.tmpdir=/tmp/elasticsearch-12548996338407439436 -XX:+HeapDumpOnOutOfMemoryError -XX:HeapDumpPath=/data0/**/log -XX:ErrorFile=/data0/**/log/hs_err_pid%p.log -Dio.netty.allocator.type=pooled -XX:MaxDirectMemorySize=16642998272 -Des.path.home=/data0/elasticsearch -Des.path.conf=/etc/elasticsearch -Des.distribution.flavor=default -Des.distribution.type=tar -Des.bundled_jdk=true -cp /data0/elasticsearch/lib/* org.elasticsearch.bootstrap.Elasticsearch -p /data0/elasticsearch/log/elasticsearch.pid --quiet
```

hot_threads
```
{***}{HpLZLh61R3KABQH6X8tT_Q}{jx2VbKTRQLOfH651XSfoIw}{***}{***:9300}{di}{rack_name=***, xpack.installed=true, tag=hot}
   Hot threads at 2020-03-26T07:03:20.599Z, interval=500ms, busiestThreads=3, ignoreIdleThreads=true:
   
   101.0% (504.8ms out of 500ms) cpu usage by thread 'elasticsearch[***][write][T#3]'
     3/10 snapshots sharing following 191 elements
       java.base@11.0.3/java.lang.ThreadLocal$ThreadLocalMap.expungeStaleEntry(ThreadLocal.java:637)
       java.base@11.0.3/java.lang.ThreadLocal$ThreadLocalMap.remove(ThreadLocal.java:519)
       java.base@11.0.3/java.lang.ThreadLocal.remove(ThreadLocal.java:242)
       java.base@11.0.3/java.util.concurrent.locks.ReentrantReadWriteLock$Sync.tryReleaseShared(ReentrantReadWriteLock.java:430)
       java.base@11.0.3/java.util.concurrent.locks.AbstractQueuedSynchronizer.releaseShared(AbstractQueuedSynchronizer.java:1382)
       java.base@11.0.3/java.util.concurrent.locks.ReentrantReadWriteLock$ReadLock.unlock(ReentrantReadWriteLock.java:897)
       app//org.elasticsearch.common.util.concurrent.ReleasableLock.close(ReleasableLock.java:49)
       app//org.elasticsearch.index.translog.Translog.readOperation(Translog.java:664)
       app//org.elasticsearch.index.engine.InternalEngine.get(InternalEngine.java:662)
       app//org.elasticsearch.index.shard.IndexShard.get(IndexShard.java:928)
       app//org.elasticsearch.index.get.ShardGetService.innerGet(ShardGetService.java:169)
       app//org.elasticsearch.index.get.ShardGetService.get(ShardGetService.java:93)
```

i cannot provide heap dump because it hava a lot of sensitive data. a screen shot is here:

![image](https://user-images.githubusercontent.com/1571034/92317487-9579b080-f033-11ea-8c04-daa4a3db162a.png)

**Comment by dnhatn** [comment]
@cztchoice Did you set the setting `index.refresh_interval` to a very small value?

**Comment by cztchoice** [comment]
here is our index settings, index.refresh_interval is the default 1s
```
{
  "index_name" : {
    "settings" : {
      "index" : {
        "mapping" : {
          "total_fields" : {
            "limit" : "5000"
          }
        },
        "refresh_interval" : "1s",
        "number_of_shards" : "180",
        "translog" : {
          "flush_threshold_size" : "4096mb",
          "sync_interval" : "60s",
          "durability" : "async"
        },
        "soft_deletes" : {
          "enabled" : "false"
        },
        "creation_date" : "1596211200114",
        "number_of_replicas" : "1",
        "uuid" : "2stsAqnbRae6iEkzVlS4-g",
        "version" : {
          "created" : "7040299"
        }
      }
    }
  }
}
```

and cluster health:
```
{
  "cluster_name" : "name",
  "status" : "green",
  "timed_out" : false,
  "number_of_nodes" : 43,
  "number_of_data_nodes" : 40,
  "active_primary_shards" : 2487,
  "active_shards" : 4843,
  "relocating_shards" : 0,
  "initializing_shards" : 0,
  "unassigned_shards" : 0,
  "delayed_unassigned_shards" : 0,
  "number_of_pending_tasks" : 0,
  "number_of_in_flight_fetch" : 0,
  "task_max_waiting_in_queue_millis" : 0,
  "active_shards_percent_as_number" : 100.0
}
```

**Comment by dnhatn** [comment]
@cztchoice Thank you for the reply. Can you share the [segment](https://www.elastic.co/guide/en/elasticsearch/reference/current/indices-segments.html) stats?

**Comment by cztchoice** [comment]
Here is the index segment info
[segment_info.json.zip](https://github.com/elastic/elasticsearch/files/5236697/segment_info.json.zip)

**Comment by dnhatn** [comment]
@cztchoice Sorry for not being clear. Can you share hot threads, segment stats, and index settings again while the issue is happening? Thank you!

**Comment by cztchoice** [comment]
sorry, we have restart the ES process to recover from this issue, so I can not get these info now.

the hot threads snippet & thread dump info is get when we have this issue.

**Comment by zhaoboc** [comment]
> @hydrogen666 Several places in Lucene also use CloseableThreadLocal. I do not want to miss these places. Can you share the heap dump privately so we can confirm these ThreadLocal are really from VersionsAndSeqNoResolver ? Thank you!

![image](https://user-images.githubusercontent.com/28853412/108712460-c3a36c00-7551-11eb-9842-1168fa780146.png)

I have a ThreadLocalMap that is half full: 11,726 WeakReference instances in a 32k table. 

Most of these weak references are empty and we can't tell what was referenced. Non-empty references are from VersionsAndSeqNoResolver as well as SegmentCoreReaders.

**Comment by zhaoboc** [comment]
Another ThreadLocalMap from [management] thread shows non-empty references are from SegmentCoreReaders mostly. 
![image](https://user-images.githubusercontent.com/28853412/108715529-f64f6380-7555-11eb-868e-9006b46194e3.png)

**Comment by jgq2008303393** [comment]
Hi, @dnhatn, is there any progress on this issue? We have the same problem.

**Comment by dnhatn** [comment]
I have unassigned myself from the issue so others can pick it up.

**Comment by jgq2008303393** [comment]
I have send a heap dump and a flame graph through [the support system](https://support.elastic.co/cases/5008X000025ZC8tQAG). We can clearly see the same problem: the thread keeps running on the expungeStaleEntry function, and the size of the write/management thread pool is very large.

We hope that Elastic can fix this  long-standing problem.

**Comment by DaveCTurner** [comment]
Thanks for the heap dump @jgq2008303393. It's very hard to make progress on this sort of issue without a heap dump, but the extra detail you provided has helped us understand the situation better. It's not something that seems to affect any of our internal benchmarks so it's a little tricky to validate possible fixes, but #85896 might address the problem. Can you benchmark this change with your workload and confirm whether it improves things?

**Comment by jgq2008303393** [comment]
> To be clear, the heap dump does not show evidence of a true leak as the issue title suggested. It's really just that some cleanup operations are delayed more than we'd like. I have adjusted the title to match.

We are agree with this. It's just some cleanup operations are delayed as we can workaround this problem by manual GC(./jmap -histo:live $pid).

As we also see large size ThreadLocalMap in management thread pool, we are afraid this is also high risk. 
![fault_cluster_management_thread (1)](https://user-images.githubusercontent.com/657140/163743195-db88893b-3fc0-4a37-a0a3-bbe177115722.jpg)

Would it be better to fix this by improving CloseableThreadLocal implemention? We're thinking about optimizing it this way：
```
public class CloseableThreadLocal<T> implements Closeable {

  // Note: remove this to avoid the ThreadLocalMap problem.
  private ThreadLocal<WeakReference<T>> t = new ThreadLocal<>();

  // Note: keep this map to emulate the ability of ThreadLocal, which allows GC when a thread exits.
  private Map<Thread,T> hardRefs = new WeakHashMap<>();

  ...
}
```

**Comment by DaveCTurner** [comment]
The `management` thread usages were removed in #75674 so it's just the `VersionsAndSeqNoResolver` usages that need addressing.

Your proposed change to `CloseableThreadLocal` introduces a need for locking on every read which seems likely to cause performance problems in some setups, noting especially that this is a change in Lucene which has many more users than just Elasticsearch. I'll raise it for discussion with the Lucene folks.

**Comment by jgq2008303393** [comment]
> Your proposed change to CloseableThreadLocal introduces a need for locking on every read which seems likely to cause performance problems in some setups, noting especially that this is a change in Lucene which has many more users than just Elasticsearch. I'll raise it for discussion with the Lucene folks.

The locking is lightweight as it just locks hardRefs when calling the get method. Looking forward to the discussion in the Lucene community.

**Comment by dnhatn** [comment]
Fixed in #93984
